Document data streams and custom index lifecycle policies

bmorelli25 opened the pull request:
### Summary

* Adds documentation for data streams
* Adds documentation for default index lifecycle polices
* Adds documentation for custom index lifecycle policies with component templates
* Moves all "storage" content under the "Manage storage" heading

### Preview

Manage storage bucket: https://apm-server_6553.docs-preview.app.elstc.co/guide/en/apm/guide/current/manage-storage.html 
Data streams: https://apm-server_6553.docs-preview.app.elstc.co/guide/en/apm/guide/current/apm-data-streams.html
ILM/ILP: https://apm-server_6553.docs-preview.app.elstc.co/guide/en/apm/guide/current/ilm-how-to.html

### Related

For https://github.com/elastic/observability-docs/issues/1073
Closes https://github.com/elastic/observability-docs/issues/1223
Closes https://github.com/elastic/observability-docs/issues/1224

bmorelli25 commented on docs/how-to.asciidoc:
```diff
@@ -4,20 +4,11 @@
 Learn how to perform common APM configuration and management tasks.
 
 * <<source-map-how-to>>
-* <<ilm-how-to>>
 * <<jaeger-integration>>
 * <<ingest-pipelines>>
-* <<manage-storage>>
-* <<apm-tune-elasticsearch>>
 
 include::./source-map-how-to.asciidoc[]
 
-include::./ilm-how-to.asciidoc[]
-
 include::./jaeger-integration.asciidoc[]
 
 include::./ingest-pipelines.asciidoc[]
-
-include::./manage-storage.asciidoc[]
-
-include::./apm-tune-elasticsearch.asciidoc[]
```
Curious what y'all think about moving this content. This PR creates a top-level bucket called, "Manage Storage":
<img width="344" alt="Screen Shot 2021-11-08 at 12 36 15 PM" src="https://user-images.githubusercontent.com/5618806/140815034-859f9456-e6dc-475c-968f-28b8933becba.png">

axw replied:
I like it. I'm not sure that "Processing and performance" belongs there though, as that's about more than just storage -- it also covers apm-server's processing overhead.

apmmachine commented:
## :grey_exclamation: Build Aborted
  
  > There is a new build on-going so the previous on-going builds have been aborted.
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-6553/detail/PR-6553/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-6553/detail/PR-6553/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-6553/detail/PR-6553/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-6553/detail/PR-6553/1//artifacts)   [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/apm-ci/transactions/view?rangeFrom=2021-11-08T20:31:08.356Z&rangeTo=2021-11-08T20:51:08.356Z&transactionName=apm-server%2Fapm-server-mbp%2FPR-%7Bnumber%7D&transactionType=unknown&latencyAggregationType=avg&traceId=daa7db198b6f0931dce3254037783eda&transactionId=173fa7591c848e83)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats



* Reason: `Aborted from #2`


* Start Time: 2021-11-08T20:41:08.356+0000


* Duration: 23 min 50 sec


* Commit: 2861fa3bd1f26313c6102c458b8e7a183daaf977



#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 1  |
| Passed       | 3340  |
| Skipped      | 10 |
| Total        | 3351   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->

  ### Test errors [![1](https://img.shields.io/badge/1%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-6553/detail/PR-6553/1//tests)
<details><summary>Expand to view the tests failures</summary><p>
    
    
##### `Build and Test / windows build-test / TestRegisterPipelines – github.com/elastic/apm-server/ingest/pipeline`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestRegisterPipelines
 
 ```
</p></details>
</ul>
    
</p></details>


<!-- STEPS ERRORS IF ANY -->











### Steps errors [![2](https://img.shields.io/badge/2%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-6553/detail/PR-6553/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Compress`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-server/pipelines/apm-server-mbp/pipelines/PR-6553/runs/1/steps/742/log/?start=0">here</a></li>
<li>Description: <code>tar --exclude=system-tests-linux-files.tgz -czf system-tests-linux-files.tgz system-tests</code></l1>
</ul>
  
  
  
  
##### `Print Message`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-server/pipelines/apm-server-mbp/pipelines/PR-6553/runs/1/steps/764/log/?start=0">here</a></li>
<li>Description: <code>[39;49m[INFO] Can not determine redirect link!!![0m</code></l1>
</ul>
  
  </p>
  </details>


## :robot: GitHub comments

To re-run your PR in the CI, just comment with:


- `/test` : Re-trigger the build.

- `/hey-apm` : Run the hey-apm benchmark.

- `/package` : Generate and publish the docker images.




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

bmorelli25 commented:
/test

axw commented on docs/ilm-how-to.asciidoc:
```diff
@@ -1,18 +1,191 @@
+To apply your new index lifecycle policy to a data stream,
+and to ensure the policy continues to be applied in future versions,
+you must create a component template.
+The component template name must start with `.`, follow the data stream naming scheme,
```
Where did you get the `.` part from? AFAIK they do not have a dot prefix.

bmorelli25 replied:
The dot (and these docs) are adapted from a tutorial that @hop-dev wrote.

hop-dev replied:
@axw Thanks for pointing this out, I think this is a mistake in my original guide, you're right the component templates don't have a dot prefix. 

@joshdover I can't see any record of us discussing having a dot prefix on the namespace specific component templates, I think we may be able to remove it? If so I'll get a PR in for the docs.

axw commented on docs/ilm-how-to.asciidoc:
```diff
@@ -1,18 +1,191 @@
+
+[discrete]
+[[data-streams-custom-four]]
+=== Step 4: Clone and modify the existing index template
```
You shouldn't need to touch any index templates for customising the ILM policy. Just edit the `@custom` component template, which is used by the Fleet-installed index template.

Did you try that and it didn't work?

bmorelli25 replied:
No, sorry, I didn't try that. I was basing this off of what @hop-dev wrote for other integrations. I can try that too.

bmorelli25 replied:
I think you're right, @axw. Looking back at the original issue, Mark pointed out that integrations that define their own ILM policy can just edit the `@custom` template: https://github.com/elastic/observability-docs/issues/986#issuecomment-914211466

bmorelli25 replied:
I've confirmed I can directly edit the `@custom` component template. But that brings up two thoughts:

1. Is this persisted on upgrade?
2. I don't think this method allows the same level of granularity that creating your own index template does. For example, when I create my own index template, I can add an ILM policy that applies only to a specific namespace, like `traces-apm-production*`. With this method, my custom template is only applied to the default index pattern (all namespaces) of `traces-apm-*`. Maybe that's okay though? IDK. That's my question for you 😄

axw replied:
> Is this persisted on upgrade? (not sure how to test this)

Yes

> I don't think this method allows the same level of granularity that creating your own index template does. For example, when I create my own index template, I can add an ILM policy that applies only to a specific namespace, like traces-apm-production*. With this method, my custom template is only applied to the default index pattern (all namespaces) of traces-apm-*. Maybe that's okay though? IDK. That's my question for you smile

Indeed, this is a good point. I'd separate it into use cases, so we don't lead users through unnecessary complexity when they want to do something simple. The basic use case is when you want to customise the ILM policy for a data type (e.g. traces); this can be done just be editing the `@custom` component template. The advanced use case is to customise the ILM policy for a data type and namespace (e.g. production); this requires introducing a new index template with higher priority.

Probably a good idea to document both, but longer term we need to make all of this less manual.

bmorelli25 commented on docs/data-streams.asciidoc:
```diff
@@ -1,4 +1,86 @@
-// to do: fill with content. placeholder for external links for now
+****
+{agent} uses data streams to store append-only time series data across multiple indices
+while giving users a single named resource for requests.
```
It just means that you can send indexing and search requests directly to a data stream -- which will route the request to backing indices that store the stream’s data.

bmorelli25 replied:
But maybe that's too confusing for these docs. I can clean this up 👍

bmorelli25 commented:
* Removed `.` prefix
* Updated the tutorial to edit the default `<data-stream-name>@custom` component template
* Added a new section that briefly explains that it is possible to have more granular, namespace-level index lifecycle policies, and links to the [tutorial in the Fleet User Guide](https://www.elastic.co/guide/en/fleet/7.16/data-streams.html#data-streams-ilm-tutorial) for more information.

bmorelli25 commented:
/test

axw reviewed (approved):
Looks great!
